suggest constraints to specify for export based on generated shape guards

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98463

The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

avikchaudhuri committed:
suggest constraints to specify for export based on generated shape guards

The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

[ghstack-poisoned]

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2231,14 +2523,16 @@ def _add_guard(self, expr: "sympy.Expr") -> None:
-                expr = LoggingShapeGuardPrinter(self.var_to_sources).doprint(expr)
-                log.warning(f"Adding shape guard {expr} at \n{user_stack}")
+                tc = TracingContext.get()
+                if tc is not None:
```
Wait why this change? TC should never be None here.

avikchaudhuri replied:
Cut. I remember running into a problem here, maybe in some other context...

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1905,14 +2190,21 @@ def hint():
                 if len(bounds) > 1:
                     exprs.append(" <= ".join(bounds))
 
+        if torch._dynamo.config.dynamic_shapes and torch._dynamo.config.summarize_dim_constraints:
+            dim_constraints.solve()
+            log.warning(f"Summary of dimension constraints:{dim_constraints.prettify_results()}")
```
No dynamo config notions in this file please. We already have dynamo source notions here in some places and it kills me.

avikchaudhuri replied:
How would you suggest I do this, then, assuming I do want to gate it with a flag?

ezyang replied:
Well, the other way we've been doing it is passing these in inside `__init__`. BTW there's no point in testing `dynamic_shapes`, you'll never hit ShapeEnv without it True.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
         super().__init__(var_to_sources, lambda n: n.name(), var_to_sources)
 
 
+class DynamicDimConstraintPrinter(StrPrinter):
```
Please document this. 

What Is the goal of this printer? Produce guard code? debug source? etc.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+                if not (s in self._substitutions) or not sympy.checksol(congruence, {s: self._substitutions[s]}):
+                    self._dynamic_results.add(self._dcp.doprint(sympy.Eq(congruence, 0)))
+
+    def prettify_univariate_inequalities(self):
```
not called?

avikchaudhuri replied:
Cut

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+            buf += f"\n\t{expr},"
+        return buf + "\n]"
+
+    def prettify_congruences(self):
```
not called?

avikchaudhuri replied:
Cut

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+            buf += f"\n\t],"
+        return buf + "\n}"
+
+    def prettify_multivariate_inequalities(self):
```
not called?

avikchaudhuri replied:
Cut

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1811,6 +2088,8 @@ def hint():
         #    if we have an input (2, 3), we must show s0*2 == 2 and s1 == 3.
         #    This does a lot of work: it covers duck sizing and equality guards.
         exprs = []
+        dim_constraints = DimConstraints(symbol_to_source, self.var_to_val)
```
This work is unused if summarize_dim_constraints is False, right?

avikchaudhuri replied:
Yeah. I thought about hiding it but it's not anything heavy that we don't do already (e.g., we're not calling any solvers when initializing).

voznesenskym commented on torch/_dynamo/config.py:
```diff
@@ -118,6 +118,10 @@
 
+# Simplify guards, summarizing static and dynamic constraints on dimensions.
+# NOTE: This only has an effect when dynamic_shapes=True.
+summarize_dim_constraints = False
```
Why is this gated, btw? If it's generally good and able, we should include it unequivocally, right? Or what is the tradeoff space for enabling this?

avikchaudhuri replied:
I'd like to leave this gated for export only, at least until we have been using it for a while. This is mainly because I have only tested this in the wild on export use cases, and I don't know (a) what performance bar we need to meet for eager (b) what corner cases we could be missing that could blow up these solvers, which we haven't used before.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+        # Our inequality solver can handle / but not %. So we need to transform them away.
+        # We do so by using the values of variables as hints to evaluate %.
+        # For soundness we record additional congruence guards and solve them separately.
+        self._var_to_val = var_to_val
```
copy to not accidentally mess with this

avikchaudhuri replied:
We're evidently not messing with it. Read-only for substitutions.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+        self._static_results = set()
+        self._dynamic_results = set()
+
+        self._dcp = DynamicDimConstraintPrinter(symbol_to_source)
```
copy symbol_to_source to not accidentally mess with this

avikchaudhuri replied:
We're not messing with it.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+                self._univariate_inequalities[s].add(expr)
+
+    def add_equality(self, source, expr):
+        if expr.free_symbols:
+            # these will resolve to either specializations or dynamic equality constraints
+            self._symbolic_equivalences.append((source, expr))
+        else:
+            # specialization, right here
+            self._static_results.add(f"{source.name()} == {expr}")
```
Wait, why only here? It seems like every expr should be checked for this, right?

avikchaudhuri replied:
Could you elaborate? We have an assertion in `add()`, we don't expect non-true non-false exprs with no free symbols.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+    def solve(self):
+        # as long as there are symbols with equalities, solve for them
+        while(self._symbols_with_equalities):
+            s = self._symbols_with_equalities.pop()
+            exprs = self._univariate_inequalities.pop(s)
+            solution = sympy.solvers.inequalities.reduce_inequalities(exprs, s)
```
Are these 2 structures enforced to be the same size somewhere?

avikchaudhuri replied:
Yes, added comment.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+        # Solutions come in two forms:
+        # - (static) specializations
+        # - (dynamic) inequalities / congruences
+        self._static_results = set()
```
what are the types here? I see you throwing strings in?

avikchaudhuri replied:
Added types in init.

voznesenskym reviewed (changes requested):
Tests, way more docs please, have a few sus spots to check out, otherwise seems like a great direction

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+            congruence = (base - mod_reduced) % divisor
+            if congruence != 0:
+                self._congruences[s].add(congruence)
+            return (base - mod_reduced) / divisor
```
Trust upstream.

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
+        return reduced_congruences
+
+    def solve(self):
+        # as long as there are symbols with equalities, solve for them
```
No, added comment

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1718,21 +1993,23 @@ def produce_guards(
             assert callable(msg_fn)
             constraint_violations.append(msg_fn)
 
+        def is_dim(src):
+            return isinstance(src, TensorPropertySource) and src.prop is TensorProperty.SIZE
```
Elsewhere in the same file we specifically have this pattern (e.g., see `track_symint`). I don't see an abstraction here that's leaking.

ezyang replied:
I'm not sure why SIZE should be treated specially here but not the other sizevars.

avikchaudhuri commented:
@voznesenskym PTAL, added more docs and tests.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1820,6 +2094,8 @@ def hint():
                 ):
                     continue
+                if is_dim(source):
+                    dim_constraints.add_equality(source, expr)
```
It's kind of annoying that you have to remember to do a dim_constraints.add anywhere you push to exprs but I can't think way of a good way to force someone to do this

ezyang replied:
Actually, can't we just get rid of expr entirely, and use dim_constraints to generate the guards in the end? Why can't we do this?

avikchaudhuri replied:
I don't see any reason not to do it, other than wanting to limit the blast radius in one go. There's some parallel simplification that we could get rid of.

ezyang replied:
OK, happy to agree this as PoR and then ship PR as is. But I don't want to turn it on "for real" without translation validation.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,274 @@ def __init__(self, var_to_sources):
+        self.symbol_to_source = symbol_to_source
+
+    def print_source(self, source) -> str:
+        return f"dynamic_dim({source.base.name()}, {source.idx})"
```
This is going to print L['x'] instead of x, I think

ezyang replied:
With export, you kind of want to do a rewrite, but not in the way you expect. Suppose you have

```
def f(x, y):
  ...

inp = [torch.randn(3), torch.randn(3)]
torch._dynamo.export(f, inp)
```

for the result to be directly actionable, you have to directly say `dynamic_dim(inp[0], 0)` e.g.

Your current message falls in an uncanny valley, where it sort of looks like it is copy pastable, but it's not. It's better to go all the way, or not at all.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1341,6 +1343,274 @@ def __init__(self, var_to_sources):
+            for congruence in congruences:
+                # any congruence that cannot be checked becomes a dynamic constraint as well
+                if not (s in self._substitutions) or not sympy.checksol(congruence, {s: self._substitutions[s]}):
+                    self._dynamic_results.add(self._dcp.doprint(sympy.Eq(congruence, 0)))
```
Are you at all willing to write a translation validator for your solving here via Z3? Based on the discussion at https://github.com/pytorch/pytorch/pull/96616#issuecomment-1484839906

This would give a lot more confidence than just the tests.

avikchaudhuri replied:
I'm up for it. The best I do here is ensure that at least the concrete values are solutions of original and simplified. But yeah, no equivalence / implication checks.

ezyang commented on test/test_dynamic_shapes.py:
```diff
@@ -816,5 +833,944 @@ def is_complex(x):
+        solution = reduce_inequalities(exprs, s).as_set()
+        self.assertEqual(solution, {8})
+
+    def test_dim_constraints_solve_full(self):
```
These constraints were lifted from somewhere, right? Seems worth mentioning where you got it from.

ezyang reviewed (approved):
I have not closely reviewed the simplification algorithm.

But I guess I don't see any reason not to merge this. The current proposed use case doesn't even need to be sound, since you'll just blast in the constraints and then we will resolve the export according to the new constraints.

voznesenskym reviewed (approved):
Thanks for writing this.

Consider taking the code from https://github.com/pytorch/pytorch/pull/94694 and writing a z3 verifier

voznesenskym reviewed (approved):
Thanks for writing this.

Consider taking the code from https://github.com/pytorch/pytorch/pull/94694 and writing a z3 verifier

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "suggest constraints to specify for export based on generated shape guards"


The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge -f "unrelated ci failures (doctr)"
